### PR TITLE
Parameterized options

### PR DESCRIPTION
**Pull request by kristianpd**
## Why?

This PR proposes the ability to provide additional information to your `Action`s and `Consideration`s via a `IParameterProvider` set on the parent `Option`.

Here is my scenario that I was unable to solve with the current `Context` alone.

- A `Ship` has many `Turret`s
- `Turret`s are not unique, meaning there can be many of the same `Turret` on the Ship
- The AI is sitting on the `Ship`, the AI needs to weigh the option of firing each `Turret` against every other `Turret`

My AI uses a `ShipContext`, which in turns contains a `Ship` reference. 

When building `Action`s and `Consideration`s for firing each `Turret`, I need the ability to specify which `Turret` to look up in the `Context`. Examples of generic (all turrets share the same consideration) considerations:

- `TurretOnCooldownConsideration`
- `TurretEnergyAvailableConsideration`
- `TurretNearDestructionConsideration`

Each of these considerations can fetch `shipContext.Ship.Turrets`, but need an additional hint to determine what `Turret`s values to use with the `Utility` evaluator. 

```c#
class TurretParameters : IParameterProvider {
  public int TurretSlot; 
}
```

Solves this by giving me the ability to 

```c#
Consider() {
  TurretParameters params = (TurretParameters)Parameters;
  Turret turret = shipContext.Ship.Turrets[params.TurretSlot];
  // turret.Cooldown
}
```

# How?

`IConsideration` and `IAction` are given a `IParameterProvider Parameters` property.  The `Option` in turn gets its `Parameters` via its `IConsideration > ICompositeConsideration > CompositeConsideration > Option` inheritance. 

When `SetAction` and `AddConsideration` are called on `Option`, the `Parameters` are cloned and set on the respective actions and considerations. They can use them by casting their `IParameterProvider Parameters` reference to the appropriate subclass of `IParameterProvider`. This has the implicit contract of the `Action` and `Consideration` being aware of what type of Parameters they use.

@ThelDoctor let me know how this looks. I'm *fairly* happy with the tests as they are now, but I suspect there may be a few areas I could change.

### CODE REVIEW

File: CrystalAI/Source/Considerations/CompositeConsideration.cs [hunk 1]
```diff
@@ -132,8 +138,9 @@ public class CompositeConsideration : ICompositeConsideration {
       if(_collection.Contains(considerationId) == false)
         return false;
 
-      InternalAddConsideration(considerationId);
-      return true;
+
+      IConsideration consideration = _collection.Create(considerationId);
+      return AddConsideration(consideration);
```
Comment (kristianpd) [inline CrystalAI/Source/Considerations/CompositeConsideration.cs:143]: I wanted to only `override` a single `AddConsideration` method, so I chose the one taking the `IConsideration itself`. This also make sure adding via `string` goes through the same path as adding via `IConsideration`

File: CrystalAI.Tests/Tests/Options/OptionParameterTests.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      Assert.That(powerOption.SetAction("a1"), Is.True);
+      Assert.That(powerOption.AddConsideration("c1"), Is.True);
+
+
+      var context = new OptionContext();
+      context.XVal9 = util;
+      linearOption.Consider(context);
+      powerOption.Consider(context);
+
+      if(util == 0.0f || util == 10.0f) {
```
Comment (kristianpd) [inline CrystalAI.Tests/Tests/Options/OptionParameterTests.cs:80]: I originally had 10 steps in this test, but it seemed redundant, so I cut it down to these 3. I could rewrite this if you think it's too confusing.

**Comment by igiagkiozis** [comment]
Hi @kristianpd thank you for the contribution. As you're probably aware of already, I'm ambivalent about adding state to the ActionBase/Consideration/Option classes. That said, this week I have a little more free time compared to last week, so I'll delve into the the code and have a look if the code integrates well in Crystal, if it needs fixing and/or if there is a better solution to the problem that motivated you adding this feature.

**Comment by kristianpd** [comment]
I'm currently reworking my AI a little, so i'm going to close this for now unless it comes back up with in my rework. Better to keep it as stateless as possible.

**Pull request closed without merging**